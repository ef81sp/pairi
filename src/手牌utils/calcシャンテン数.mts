import { flattenRest } from "./seek有効牌.mjs"
import { countRequiredブロックnum } from "./utils/countRequiredブロックnum.mjs"
import { count牌 } from "./utils/count牌.mjs"
import { is暗刻 } from "./utils/is暗刻.mjs"
import type {
  ExtractResult5ブロック,
  ExtractResult七対子,
  ExtractResult国士無双,
} from "./手牌utils.type.mjs"

export const calcシャンテン数5ブロック = (
  extractResult5ブロック: ExtractResult5ブロック,
): number => {
  const requiredブロックnum = countRequiredブロックnum(count牌(extractResult5ブロック))

  const required面子num = requiredブロックnum - 1
  const { 雀頭, 面子, 塔子, rest: _rest } = extractResult5ブロック
  const rest = flattenRest(_rest)

  // いわゆる「8点法」を採用
  let シャンテン数 = 8

  // 手牌が短い場合、必要面子数が減る
  シャンテン数 -= (5 - requiredブロックnum) * 2

  // 面子の数 * 2点
  シャンテン数 -= 面子.length * 2
  // 塔子の数 * 1点
  シャンテン数 -= 塔子.length
  // 雀頭があれば 1点
  シャンテン数 -= 雀頭 ? 1 : 0

  // 例外パターン
  // 面子と塔子が5個以上あると +1点 (雀頭は関係ない)
  if (面子.length + 塔子.length >= requiredブロックnum) {
    シャンテン数 += 1
  }
  // 雀頭と塔子が同一(4枚使い)だと +1点
  if (
    雀頭 &&
    塔子.some(
      (塔子) =>
        塔子.component[0].toEqual(塔子.component[1]) &&
        塔子.component[0].toEqual(雀頭.component[0]),
    )
  ) {
    シャンテン数 += 1
  }
  // ヘッドレスで浮き牌が暗刻で使われていると +1点
  if (!雀頭 && 面子.length + 塔子.length === required面子num) {
    const 暗刻List = 面子.filter((m) => is暗刻(m))
    if (rest.every((p) => 暗刻List.some((面子) => 面子.component[0].toEqual(p)))) {
      シャンテン数 += 1
    }
  }
  return シャンテン数
}
export const calcシャンテン数七対子 = (extractResult七対子: ExtractResult七対子): number => {
  let result = 6 - extractResult七対子.対子.length

  // 手牌の牌の種類が7種類未満の場合、足りない分だけシャンテン数が増える
  const uniqueFlattenRest = flattenRest(extractResult七対子.rest)
    .filter((p, index, array) => array.findIndex((p2) => p2.toEqual(p)) === index)
    .map((p) => p.toString())

  const unique牌List = Array.from(
    new Set([
      ...extractResult七対子.対子.map((対子) => 対子.component[0].toString()),
      ...uniqueFlattenRest,
    ]),
  )
  const kinds牌 = unique牌List.length
  if (kinds牌 < 7) {
    result += 7 - kinds牌
  }

  return result
}

export const calcシャンテン数国士無双 = (extractResult国士無双: ExtractResult国士無双): number => {
  return extractResult国士無双.么九牌rest.length === 0
    ? 13 - extractResult国士無双.么九牌unique.length
    : 12 - extractResult国士無双.么九牌unique.length
}
