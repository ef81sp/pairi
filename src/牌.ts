import { Number字牌, Number数牌, Str牌, Suit, Suit字牌, Suit数牌 } from "./utils/types"

/**
 * Represents a 牌 (Mahjong tile).
 */
export class 牌 {
  readonly suit: Suit
  readonly number: typeof this.suit extends Suit字牌 ? Number字牌 : Number数牌
  readonly is赤牌: boolean
  private readonly str牌: Str牌

  /**
   * Constructs a new 牌 instance.
   * @param pai - The string representation of the 牌.
   * @param isRed - Indicates whether the 牌 is a red tile.
   */
  constructor(pai: Str牌, isRed = false) {
    const { suit, number } = parsePaiString(pai)
    this.suit = suit
    this.number = number
    this.is赤牌 = (() => {
      if (suit === "z") return false
      return number === 5 && isRed
    })()
    this.str牌 = pai
  }

  /**
   * Returns the string representation of the 牌.
   * @returns The string representation of the 牌.
   */
  toString() {
    return this.str牌
  }

  /**
   * Checks if the 牌 is equal to another 牌.
   * @param pai - The 牌 to compare.
   * @returns `true` if the 牌 is equal to the specified 牌, `false` otherwise.
   */
  toEqual(pai: 牌) {
    return this.str牌 === pai.str牌
  }

  /**
   * Creates a clone of the 牌.
   * @returns A new 牌 instance that is a clone of the current 牌.
   */
  clone() {
    return new 牌(this.str牌, this.is赤牌)
  }
}

/**
 * Parses a string representation of a 牌 and returns the corresponding suit and number.
 * @param pai - The string representation of the 牌.
 * @returns An object containing the suit and number of the 牌.
 * @throws Error if the string representation is invalid.
 */
const parsePaiString = (
  pai: string,
): { suit: Suit数牌; number: Number数牌 } | { suit: Suit字牌; number: Number字牌 } => {
  if (pai.length !== 2) throw new Error(`invalid signature ${pai}: length must be 2.`)

  const number = Number(pai[0])
  const suit = pai[1]
  const numberErrorMessage = `invalid signature ${pai}: number must be 1, 2, 3, 4, 5, 6, 7, 8 or 9.`
  const numberErrorMessage数牌 = `invalid signature ${pai}: number must be 1, 2, 3, 4, 5, 6, 7, 8 or 9.`
  const numberErrorMessage字牌 = `invalid signature ${pai}: number must be 1, 2, 3, 4, 5, 6 or 7.`

  if (!Number.isInteger(number)) throw new Error(numberErrorMessage)
  if (suit === undefined || !["m", "p", "s", "z"].includes(suit))
    throw new Error(`invalid signature ${pai}: suit must be "m", "p", "s" or "z".`)

  if (["m", "p", "s"].includes(suit)) {
    if (!(1 <= number && number <= 9)) throw new Error(numberErrorMessage数牌)
    return {
      suit,
      number,
    } as { suit: Suit数牌; number: Number数牌 }
  }
  if (!(1 <= number && number <= 7)) throw new Error(numberErrorMessage字牌)
  return {
    suit,
    number,
  } as { suit: Suit字牌; number: Number字牌 }
}

/**
 * Checks if a string is a valid 牌.
 * @param str - The string to check.
 * @returns `true` if the string is a valid 牌, `false` otherwise.
 */
export const isStr牌 = (str: string): str is Str牌 => {
  if (str.length !== 2) return false
  const number = Number(str[0])
  const suit = str[1]
  if (!Number.isInteger(number)) return false
  if (suit === undefined) return false
  if (!["m", "p", "s", "z"].includes(suit)) return false
  if (["m", "p", "s"].includes(suit)) {
    if (!(1 <= number && number <= 9)) return false
  } else {
    if (!(1 <= number && number <= 7)) return false
  }
  return true
}
