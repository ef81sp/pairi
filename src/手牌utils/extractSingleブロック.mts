import { Suit } from "../utils/types.mjs"
import { isStr牌, 牌 } from "../牌.mjs"
import {
  Result単体抽出,
  Result単体抽出Success,
  Tブロック,
  TブロックType,
  T塔子,
  T対子,
  T手牌Suit別,
  T雀頭,
  T面子,
} from "./手牌utils.type.mjs"

// 用語: 順塔子とは、 両面・辺張・嵌張 のこと。一般名称がないので……

// TS、返り値の型の絞り込みは未実装らしい。asで型アサーションするしかない。

export const extractSingleブロック = <T extends TブロックType>(
  手牌: T手牌Suit別,
  type: T,
  startAt: { suit: Suit; index: number },
): Result単体抽出<T> => {
  validate手牌(手牌)

  const { suit, index } = startAt
  switch (type) {
    case "雀頭": {
      const 牌1 = 手牌[suit][index]
      const 牌2 = 手牌[suit][index + 1]
      if (!(牌1 && 牌2)) return { status: "failure" }
      if (!is対子(牌1, 牌2)) return { status: "failure" }
      const ブロック: T雀頭 = {
        type: "雀頭",
        component: [牌1, 牌2],
      }
      const result: Result単体抽出<"雀頭"> = {
        status: "success",
        ブロック,
        rest: {
          ...手牌,
          [suit]: 手牌[suit].filter((_, i) => i !== index && i !== index + 1),
        },
      }
      return result satisfies Result単体抽出Success<"雀頭"> as Result単体抽出<T>
    }
    case "面子": {
      // 刻子 → 順子
      const 牌1 = 手牌[suit][index]
      const 牌2 = 手牌[suit][index + 1]
      const 牌3 = 手牌[suit][index + 2]
      if (!(牌1 && 牌2 && 牌3)) return { status: "failure" }
      if (is刻子(牌1, 牌2, 牌3)) {
        const ブロック: T面子 = {
          type: "面子",
          component: [牌1, 牌2, 牌3],
        }
        const result: Result単体抽出<"面子"> = {
          status: "success",
          ブロック,
          rest: {
            ...手牌,
            [suit]: 手牌[suit].filter((_, i) => i !== index && i !== index + 1 && i !== index + 2),
          },
        }
        return result satisfies Result単体抽出Success<"面子"> as Result単体抽出<T>
      }
      return find順子(手牌, startAt) as Result単体抽出<T>
    }
    case "塔子": {
      // 対子 → 順塔子
      const 牌1 = 手牌[suit][index]
      const 牌2 = 手牌[suit][index + 1]
      if (!(牌1 && 牌2)) return { status: "failure" }
      if (is対子(牌1, 牌2)) {
        const ブロック: T塔子 = {
          type: "塔子",
          component: [牌1, 牌2],
        }
        const result: Result単体抽出<"塔子"> = {
          status: "success",
          ブロック,
          rest: {
            ...手牌,
            [suit]: 手牌[suit].filter((_, i) => i !== index && i !== index + 1),
          },
        }
        return result satisfies Result単体抽出Success<"塔子"> as Result単体抽出<T>
      }
      return find順塔子(手牌, startAt) as Result単体抽出<T>
    }
    case "対子": {
      // 雀頭と同じだが七対子用として分ける
      const 牌1 = 手牌[suit][index]
      const 牌2 = 手牌[suit][index + 1]
      if (!(牌1 && 牌2)) return { status: "failure" }
      if (!is対子(牌1, 牌2)) return { status: "failure" }
      const ブロック: T対子 = {
        type: "対子",
        component: [牌1, 牌2],
      }
      const result: Result単体抽出<"対子"> = {
        status: "success",
        ブロック,
        rest: {
          ...手牌,
          [suit]: 手牌[suit].filter((_, i) => i !== index && i !== index + 1),
        },
      }
      return result satisfies Result単体抽出Success<"対子"> as Result単体抽出<T>
    }
    default:
      throw new Error(`invalid type: ${type}`)
  }
}

const validate手牌 = (手牌: T手牌Suit別): void => {
  if (!手牌.m.every((p) => p.suit === "m")) {
    throw new Error(`invalid type in "手牌.m". received: ${手牌.m.map((p) => p.toString()).join()}`)
  }
  if (!手牌.p.every((p) => p.suit === "p")) {
    throw new Error(`invalid type in "手牌.p". received: ${手牌.p.map((p) => p.toString()).join()}`)
  }
  if (!手牌.s.every((p) => p.suit === "s")) {
    throw new Error(`invalid type in "手牌.s". received: ${手牌.s.map((p) => p.toString()).join()}`)
  }
  if (!手牌.z.every((p) => p.suit === "z")) {
    throw new Error(`invalid type in "手牌.z". received: ${手牌.z.map((p) => p.toString()).join()}`)
  }
  if (!手牌.m.every((p) => p.number >= 1 && p.number <= 9)) {
    throw new Error(
      `invalid number in "手牌.m". received: ${手牌.m.map((p) => p.toString()).join()}`,
    )
  }
  if (!手牌.p.every((p) => p.number >= 1 && p.number <= 9)) {
    throw new Error(
      `invalid number in "手牌.p". received: ${手牌.p.map((p) => p.toString()).join()}`,
    )
  }
  if (!手牌.s.every((p) => p.number >= 1 && p.number <= 9)) {
    throw new Error(
      `invalid number in "手牌.s". received: ${手牌.s.map((p) => p.toString()).join()}`,
    )
  }
  if (!手牌.z.every((p) => p.number >= 1 && p.number <= 7)) {
    throw new Error(
      `invalid number in "手牌.z". received: ${手牌.z.map((p) => p.toString()).join()}`,
    )
  }
}

const is対子 = (牌1: 牌, 牌2: 牌): boolean => {
  return 牌1.toEqual(牌2)
}

const is刻子 = (牌1: 牌, 牌2: 牌, 牌3: 牌): boolean => {
  return (
    牌1.suit === 牌2.suit &&
    牌2.suit === 牌3.suit &&
    牌1.number === 牌2.number &&
    牌2.number === 牌3.number
  )
}

const is順塔子 = (牌1: 牌, 牌2: 牌): boolean => {
  return 牌1.suit === 牌2.suit && 牌1.number + 1 === 牌2.number
}

const find順子 = (
  手牌: T手牌Suit別,
  startAt: { suit: Suit; index: number },
): Result単体抽出<"面子"> => {
  const 手牌suit = 手牌[startAt.suit]
  const start牌 = 手牌suit[startAt.index]

  if (start牌 === undefined) return { status: "failure" }
  if (start牌.suit === "z") return { status: "failure" }

  // 8や9から順子は作れない
  if (start牌.number === 8 || start牌.number === 9) return { status: "failure" }

  const 牌2str = `${start牌.number + 1}${start牌.suit}`
  const 牌3str = `${start牌.number + 2}${start牌.suit}`

  if (!(isStr牌(牌2str) && isStr牌(牌3str))) return { status: "failure" }

  const 順子: T面子 = {
    type: "面子",
    component: [start牌, new 牌(牌2str), new 牌(牌3str)],
  }

  const 牌2index = 手牌suit.findIndex((p) => p.toString() === 牌2str)
  const 牌3index = 手牌suit.findIndex((p) => p.toString() === 牌3str)

  if (牌2index === -1 || 牌3index === -1) return { status: "failure" }
  return {
    status: "success",
    ブロック: 順子,
    rest: {
      ...手牌,
      [startAt.suit]: 手牌suit.filter(
        (_, i) => i !== startAt.index && i !== 牌2index && i !== 牌3index,
      ),
    },
  }
}

const find順塔子 = (
  手牌: T手牌Suit別,
  startAt: { suit: Suit; index: number },
): Result単体抽出<"塔子"> => {
  const 手牌suit = 手牌[startAt.suit]
  const start牌 = 手牌suit[startAt.index]

  if (start牌 === undefined) return { status: "failure" }
  if (start牌.suit === "z") return { status: "failure" }

  // 9から順塔子は作れない
  if (start牌.number === 9) return { status: "failure" }

  // 両面・辺張(連続)を試行
  const 牌2str = `${start牌.number + 1}${start牌.suit}`
  if (!isStr牌(牌2str)) return { status: "failure" }

  const 牌2index = 手牌suit.findIndex((p) => p.toString() === 牌2str)
  if (牌2index >= 0) {
    const 塔子: T塔子 = {
      type: "塔子",
      component: [start牌, new 牌(牌2str)],
    }
    return {
      status: "success",
      ブロック: 塔子,
      rest: {
        ...手牌,
        [startAt.suit]: 手牌suit.filter((_, i) => i !== startAt.index && i !== 牌2index),
      },
    }
  }

  // 嵌張(1個とばし)を試行
  const 牌3str = `${start牌.number + 2}${start牌.suit}`
  if (!isStr牌(牌3str)) return { status: "failure" }

  const 牌3index = 手牌suit.findIndex((p) => p.toString() === 牌3str)
  if (牌3index >= 0) {
    const 塔子: T塔子 = {
      type: "塔子",
      component: [start牌, new 牌(牌3str)],
    }
    return {
      status: "success",
      ブロック: 塔子,
      rest: {
        ...手牌,
        [startAt.suit]: 手牌suit.filter((_, i) => i !== startAt.index && i !== 牌3index),
      },
    }
  }
  return { status: "failure" }
}
