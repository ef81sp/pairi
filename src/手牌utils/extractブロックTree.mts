import { extractSingleブロック } from "./extractSingleブロック.mjs"
import {
  type IntermediateExtractResult,
  type IntermediateExtractTree,
  type TブロックType,
  type T手牌Suit別,
  T面子,
} from "./手牌utils.type.mjs"

export const extract面子Tree = (手牌: T手牌Suit別): IntermediateExtractTree<"面子">[] => {
  const wholeResult: ReturnType<typeof extract面子Tree> = []
  for (const suit of ["m", "p", "s", "z"] as const) {
    const 手牌suit = 手牌[suit]
    for (let i = 0; i < 手牌suit.length; i++) {
      const result面子 = extractSingleブロック(手牌, "面子", { suit: suit, index: i })
      if (result面子.status === "failure") continue

      const children = extract面子Tree(result面子.rest)

      wholeResult.push({
        ブロック: result面子.ブロック,
        rest: result面子.rest,
        children,
      })
    }
  }

  return wholeResult
}

export const extract塔子Tree = (手牌: T手牌Suit別): IntermediateExtractTree<"塔子">[] => {
  const wholeResult: ReturnType<typeof extract塔子Tree> = []
  for (const suit of ["m", "p", "s", "z"] as const) {
    const 手牌suit = 手牌[suit]
    for (let i = 0; i < 手牌suit.length; i++) {
      const result塔子 = extractSingleブロック(手牌, "塔子", { suit: suit, index: i })
      if (result塔子.status === "failure") {
        continue
      }

      const children = extract塔子Tree(result塔子.rest)

      wholeResult.push({
        ブロック: result塔子.ブロック,
        rest: result塔子.rest,
        children,
      })
    }
  }

  // 塔子がない場合も、restを返さないといけない
  if (wholeResult.length === 0) {
    return [
      {
        ブロック: null,
        rest: 手牌,
        children: [],
      },
    ]
  }

  return wholeResult
}

export const extract対子Tree = (手牌: T手牌Suit別): IntermediateExtractTree<"対子">[] => {
  const wholeResult: ReturnType<typeof extract対子Tree> = []
  for (const suit of ["m", "p", "s", "z"] as const) {
    const 手牌suit = 手牌[suit]
    for (let i = 0; i < 手牌suit.length; i++) {
      const result対子 = extractSingleブロック(手牌, "対子", { suit: suit, index: i })
      if (result対子.status === "failure") {
        continue
      }

      const children = extract対子Tree(result対子.rest)

      wholeResult.push({
        ブロック: result対子.ブロック,
        rest: result対子.rest,
        children,
      })
    }
  }

  // 塔子がない場合も、restを返さないといけない
  if (wholeResult.length === 0) {
    return [
      {
        ブロック: null,
        rest: 手牌,
        children: [],
      },
    ]
  }

  return wholeResult
}

export const flatTree = <T extends TブロックType>(
  tree: IntermediateExtractTree<T>,
): IntermediateExtractResult<T>[] => {
  if (tree.children.length === 0) {
    return [
      {
        ブロック: tree.ブロック ? [tree.ブロック] : [],
        rest: tree.rest,
      },
    ]
  }

  const flatChildren = tree.children.flatMap(flatTree)

  const result = flatChildren.map((c) => ({
    ブロック: (tree.ブロック ? [tree.ブロック, ...c.ブロック] : [...c.ブロック]).sort((a, b) => {
      if (a == null) return 1
      if (b == null) return -1
      // component は、suit が m > p > s > z の順、そして数字が小さい順
      if (a.type === b.type) {
        if (a.component[0].suit === b.component[0].suit) {
          return a.component[0].number - b.component[0].number
        }
        return a.component[0].suit.charCodeAt(0) - b.component[0].suit.charCodeAt(0)
      }
      // type は、雀頭 > 面子 > 塔子 の順
      if (a.type === "雀頭") return -1
      if (b.type === "雀頭") return 1
      if (a.type === "面子") return -1
      if (b.type === "面子") return 1
      if (a.type === "塔子") return -1
      if (b.type === "塔子") return 1
      return 0
    }),
    rest: c.rest,
  }))

  return result
}

export const flatTrees = <T extends TブロックType>(
  trees: IntermediateExtractTree<T>[],
): IntermediateExtractResult<T>[] => {
  return trees.flatMap(flatTree).filter((r, i, arr) => {
    const rStr = JSON.stringify(r)
    return arr.findIndex((r2) => JSON.stringify(r2) === rStr) === i
  })
}
