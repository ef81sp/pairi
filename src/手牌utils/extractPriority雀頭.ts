import { 牌 } from "../牌"
import { calcシャンテン数5ブロック } from "./calcシャンテン数"
import { extractSingleブロック } from "./extractSingleブロック"
import { extract塔子Tree, extract面子Tree, flatTrees } from "./extractブロックTree"
import {
  ExtractResult5ブロック,
  Result単体抽出,
  Result単体抽出Success,
  T手牌Suit別,
} from "./手牌utils.type"

export const extractPriority雀頭 = (手牌: T手牌Suit別): ExtractResult5ブロック[] => {
  const results: ExtractResult5ブロック[] = []
  const 雀頭List = extract雀頭(手牌)

  // ブロック(面子・塔子)を抽出する
  for (const 雀頭 of 雀頭List) {
    const 面子List = flatTrees(extract面子Tree(雀頭.rest))
    for (const 面子 of 面子List) {
      const 塔子List = flatTrees(extract塔子Tree(面子.rest))
      for (const 塔子 of 塔子List) {
        results.push({
          雀頭: 雀頭.ブロック,
          面子: 面子.ブロック,
          塔子: 塔子.ブロック,
          rest: 塔子.rest,
        })
      }
    }
  }

  // シャンテン数の低いもののみを残す
  const シャンテン数付きresult: [number, ExtractResult5ブロック][] = results.map((r) => [
    calcシャンテン数5ブロック(r),
    r,
  ])
  const min = Math.min(...シャンテン数付きresult.map(([s]) => s))
  const filteredResults = シャンテン数付きresult
    .filter(([シャンテン数]) => シャンテン数 === min)
    .map(([, r]) => r)

  return filteredResults
}

const extract雀頭 = (手牌: T手牌Suit別): Result単体抽出Success<"雀頭">[] => {
  const wholeResult: Result単体抽出<"雀頭">[] = []
  for (const suit of ["m", "p", "s", "z"] as const) {
    const 手牌suit = 手牌[suit]
    for (let i = 0; i < 手牌suit.length; i++) {
      wholeResult.push(extractSingleブロック(手牌, "雀頭", { suit: suit, index: i }))
    }
  }
  // failureを除外
  const successResult = wholeResult.filter(
    (r) => r.status === "success",
  ) as Result単体抽出Success<"雀頭">[]
  // 重複除去
  const uniqueResult = successResult.filter((r, i) => {
    const rStr = JSON.stringify(r)
    return successResult.findIndex((r2) => JSON.stringify(r2) === rStr) === i
  })
  return uniqueResult
}

if (import.meta.vitest) {
  test("雀頭抽出", () => {
    const result = extract雀頭({
      m: (["1m", "1m", "1m", "2m", "3m"] as const).map((s) => new 牌(s)),
      p: (["4p", "5p", "6p", "6p", "9p"] as const).map((s) => new 牌(s)),
      s: (["3s"] as const).map((s) => new 牌(s)),
      z: (["1z", "1z"] as const).map((s) => new 牌(s)),
    })
    expect(result).toMatchSnapshot()
  })
}
