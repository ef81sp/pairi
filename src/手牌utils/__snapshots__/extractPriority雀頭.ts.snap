// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`雀頭抽出 1`] = `
[
  {
    "rest": {
      "m": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 2,
          "str牌": "2m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 3,
          "str牌": "3m",
          "suit": "m",
        },
      ],
      "p": [
        牌 {
          "is赤牌": false,
          "number": 4,
          "str牌": "4p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 5,
          "str牌": "5p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 6,
          "str牌": "6p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 6,
          "str牌": "6p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 9,
          "str牌": "9p",
          "suit": "p",
        },
      ],
      "s": [
        牌 {
          "is赤牌": false,
          "number": 3,
          "str牌": "3s",
          "suit": "s",
        },
      ],
      "z": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1z",
          "suit": "z",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1z",
          "suit": "z",
        },
      ],
    },
    "status": "success",
    "ブロック": {
      "component": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
      ],
      "type": "雀頭",
    },
  },
  {
    "rest": {
      "m": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 2,
          "str牌": "2m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 3,
          "str牌": "3m",
          "suit": "m",
        },
      ],
      "p": [
        牌 {
          "is赤牌": false,
          "number": 4,
          "str牌": "4p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 5,
          "str牌": "5p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 9,
          "str牌": "9p",
          "suit": "p",
        },
      ],
      "s": [
        牌 {
          "is赤牌": false,
          "number": 3,
          "str牌": "3s",
          "suit": "s",
        },
      ],
      "z": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1z",
          "suit": "z",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1z",
          "suit": "z",
        },
      ],
    },
    "status": "success",
    "ブロック": {
      "component": [
        牌 {
          "is赤牌": false,
          "number": 6,
          "str牌": "6p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 6,
          "str牌": "6p",
          "suit": "p",
        },
      ],
      "type": "雀頭",
    },
  },
  {
    "rest": {
      "m": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 2,
          "str牌": "2m",
          "suit": "m",
        },
        牌 {
          "is赤牌": false,
          "number": 3,
          "str牌": "3m",
          "suit": "m",
        },
      ],
      "p": [
        牌 {
          "is赤牌": false,
          "number": 4,
          "str牌": "4p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 5,
          "str牌": "5p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 6,
          "str牌": "6p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 6,
          "str牌": "6p",
          "suit": "p",
        },
        牌 {
          "is赤牌": false,
          "number": 9,
          "str牌": "9p",
          "suit": "p",
        },
      ],
      "s": [
        牌 {
          "is赤牌": false,
          "number": 3,
          "str牌": "3s",
          "suit": "s",
        },
      ],
      "z": [],
    },
    "status": "success",
    "ブロック": {
      "component": [
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1z",
          "suit": "z",
        },
        牌 {
          "is赤牌": false,
          "number": 1,
          "str牌": "1z",
          "suit": "z",
        },
      ],
      "type": "雀頭",
    },
  },
]
`;
